libs/Microsoft.MixedReality.WebRTC/AudioTransceiver.cs

```diff
@@ -0,0 +1,146 @@
+        }
+
+        /// <summary>
+        /// Remote audio track associated with the transceiver and receiving some audio from the remote peer.
+        /// This property is updated when the transceiver negotiates a new media direction. If the new direction
+        /// allows receiving some audio, a new <see cref="RemoteVideoTrack"/> is created and assigned to the
+        /// property. Otherwise the property is cleared to <c>null</c>.
+        /// </summary>
+        public RemoteAudioTrack RemoteTrack { get; private set; } = null;
+
+        /// <summary>
+        /// Backing field for <see cref="LocalTrack"/>.
+        /// </summary>
+        private LocalAudioTrack _localTrack = null;
+
+        internal AudioTransceiver(IntPtr handle, PeerConnection peerConnection, int mlineIndex, string name, string[] streamIDs, Direction initialDesiredDirection)
+            : base(handle, MediaKind.Audio, peerConnection, mlineIndex, name, streamIDs, initialDesiredDirection)
+        {
+        }
+
+        /// <summary>
+        /// Change the local audio track sending data to the remote peer.
+        /// 
+        /// This detaches the previous local audio track if any, and attaches the new one instead.
+        /// Note that the transceiver will only send some audio data to the remote peer if its
+        /// negotiated direction includes sending some data and it has an attached local track to
+        /// produce this data.
+        /// 
+        /// This change is transparent to the session, and does not trigger any renegotiation.
+        /// </summary>
+        /// <param name="track">The new local audio track attached to the transceiver, and used to
+        /// produce audio data to send to the remote peer if the transceiver is sending.
+        /// Passing <c>null</c> is allowed, and will detach the current track if any.</param>
+        public void SetLocalTrack(LocalAudioTrack track)
+        {
+            if (track == _localTrack)
+            {
+                return;
+            }
+
+            if (track != null)
+            {
+                if ((track.PeerConnection != null) && (track.PeerConnection != PeerConnection))
+                {
+                    throw new InvalidOperationException($"Cannot set track {track} of peer connection {track.PeerConnection} on audio transceiver {this} of different peer connection {PeerConnection}.");
+                }
+                var res = TransceiverInterop.Transceiver_SetLocalAudioTrack(_nativeHandle, track._nativeHandle);
+                Utils.ThrowOnErrorCode(res);
+            }
+            else
+            {
+                // Note: Cannot pass null for SafeHandle parameter value (ArgumentNullException)
+                var res = TransceiverInterop.Transceiver_SetLocalAudioTrack(_nativeHandle, new LocalAudioTrackHandle());
+                Utils.ThrowOnErrorCode(res);
+            }
+
+            // Capture peer connection; it gets reset during track manipulation below
```

Comment:
It's not clear how this is possible, and anyway it sounds like an anti-pattern that A.x() calls B.y() that in turn changes some member in A - the dependencies are difficult to understand. This should be simplified - I'd try to suggest a way but I am not sure how this works yet 😄

Reply:
There must be a global state change from good known state A to good known state B, which involves touching multiple objects. So to avoid A=>B=>A=>B=>... recursive changes (classical SetParent/AddChild issue), there needs to be a "hierarchy":
- Any call is redirected to tracks
- Tracks self-update, then update their transceivers
- Transceivers self-update, then update their peer connection
- The chain of update stops there.

Here the transceiver receives a user call so redirect to the track (_e.g._ `parent.AddChild(child)` => `child.SetParent(parent)`). Then the track self-update, update its transceiver (the caller), and finally the peer connection. So `Transceiver.PeerConnection` gets updated by `OnTrackRemoved()` and `OnTrackAdded()`.

Reply:
Why can't it be:
- transceiver receives call
- transceiver updates track
- transceiver updates peer connection
?

A chain of side effects is much more complicated to follow/debug than a flat list.

Reply:
Because the starting point is not always a user call to the transceiver. The transceiver might be created by the implementation. Also other changes, _e.g._ to tracks, might do the same operation than the one done here as a result of changing the transceiver, so those `OnXxx` calls might be called from multiple places.

I will try to have a pass to flatten/simplify, see what we can improve.

Reply:
> Because the starting point is not always a user call to the transceiver

I don't think it matters too much that it is a _user_ call - the point is, it's a transceiver operation so some code is calling a method in the transceiver.

> those OnXxx calls might be called from multiple places

These seem to be called only here though 😄 In general, if we do want to trigger a set of side effects across multiple objects from multiple places then a chain of calls might be an option, but it makes the code more complicated and less transparent, so I'd avoid it if possible.

Reply:
Fixed.

Reply:
Much nicer now IMHO 😉